benchmarks/conda_lock_plugin.py

```diff
@@ -0,0 +1,48 @@
+"""
+ASV plug-in providing an alternative ``Environment`` subclass, which uses Nox
+for environment management.
+
+"""
+from asv.console import log
+from asv.plugins.conda import Conda, _find_conda
+
+class CondaLock(Conda):
+    """
+    Create the environment based on a **version-controlled** lockfile.
+
+    Creating the environment instance is deferred until ``install_project`` time,
+    when the commit hash etc is known and we can access the lock file.
+    The environment is then overwritten by the specification provided at the
+    ``config.conda_lockfile`` path.  ``conda.conda_lockfile`` must point to
+    an @EXPLICIT conda manifest, e.g. the output of either the ``conda-lock`` tool,
+    or ``conda list --explicit``.
+    """
+    tool_name = "conda-lock"
+
+    def __init__(self, conf, python, requirements):
+        self._lockfile_path = conf.conda_lockfile
+        super().__init__(conf, python, requirements)
+
+    def _uninstall_project(self):
+        if self._get_installed_commit_hash():
+            # we can only run the uninstall command if an environment has already
+            # been made before, otherwise there is no python to use to uninstall
+            super()._uninstall_project()
+            # TODO: we probably want to conda uninstall all the packages too
+            #       something like:
+            #       conda list --no-pip | sed /^#/d | cut -f 1 -d " " | xargs conda uninstall
+
+    def _setup(self):
+        # create the shell of a conda environment, that includes no packages
+        log.info("Creating conda environment for {0}".format(self.name))
+        self.run_executable(_find_conda(), ['create', "-y", '-p', self._path, '--force'])
```

Comment:
That is true, but I'm not sure that is a valid use case for this kind of environment.  a conda-lock file is explicit about which version of numpy it uses, it can't be easily changed out without re-resolving the whole environment.

I suppose the way to do that kind of testing would be to have a list of lockfiles which pin different sets of dependencies... but that seems like it's better served by the exisiting Conda or VirtualEnv approaches?

Reply:
My idea would be to generate the environment from a lock file, then if any `matrix` is in `asv.conf.json`: install the required package as a second step.

[Here is an example](https://github.com/SciTools-incubator/iris-esmf-regrid/blob/ee8fdecfc8633208cab7c133dec98cef6a57a4d8/benchmarks/nox_asv_plugin.py#L210), which I _think_ I tested while writing it. Certainly wouldn't be the only way of achieving the idea.